lib/eslint/eslint-helpers.js

```diff
@@ -162,16 +162,32 @@ async function findFiles({
                             return false;
                         }
 
+                        // patterns starting with ** always apply
+                        if (filePattern.startsWith("**")) {
+                            return true;
+                        }
+
                         // not sure how to handle negated patterns yet
                         if (filePattern.startsWith("!")) {
                             return false;
                         }
 
-                        // check if the pattern would be inside the cwd or not
+                        // check if the pattern would be inside the config base path or not
                         const fullFilePattern = path.join(cwd, filePattern);
-                        const relativeFilePattern = path.relative(configs.basePath, fullFilePattern);
+                        const patternRelativeToConfigBasePath = path.relative(configs.basePath, fullFilePattern);
+
+                        if (patternRelativeToConfigBasePath.startsWith("..")) {
+                            return false;
+                        }
+
+                        // check if the pattern is inside the directory or not
+                        const patternRelativeToFilePath = path.relative(filePath, fullFilePattern);
+
+                        if (patternRelativeToFilePath.startsWith("..")) {
+                            return false;
+                        }
```

Comment:
It also doesn't work when the pattern starts with a non-literal (except for the special-cased `**`).

For example, if the config entry has `files: ["{foo,bar}/*.md"]`, command `eslint foo` won't find `.md` files in `foo` because `patternRelativeToFilePath` is `..\{foo,bar}\*.md`, so the pattern gets filtered out here.

Reply:
Hmmm, yeah, this is the trickiest part of all of this. Do you have any suggestions?

Reply:
For this step (filtering out patterns that can't match inside the given directory), I think we could use minimatch with [`partial: true`](https://github.com/isaacs/minimatch#partial). But then there's the next step where we should adjust the remaining patterns to match only inside the given directory, which seems overly complex.

Reply:
Some alternative approaches:

1. Use `globby` to simply get all files in the directory (if the given directory is `foo/bar`, pass `foo/bar/**`) and then filter out those that don't match at least one pattern that doesn't end with a `*`. I think this solution is the least error-prone.
2. [Negative patterns](https://github.com/mrmlnc/fast-glob#how-to-exclude-directory-from-reading) work as logical AND, so it might be possible to get the intersection by appending some form of double negation to the list of patterns.

Reply:
Thanks for the suggestions. These were super helpful. I went with a `minimatch`-based approach that seems to work. I didn't need to use `partial: true` because we actually do have the full path at that point.